File: ui/app/lib/workflow-submission-permissions.ts
import { WorkflowType } from "./workflow-governance";

const IT_SUBSCRIPTION_TYPES: WorkflowType[] = [
  "new_subscription_request",
  "renewal_request",
];

const IT_LICENCE_TYPES: WorkflowType[] = ["license_assignment_request"];

const HR_LICENCE_TYPES: WorkflowType[] = ["license_assignment_request", "hr_onboarding_request"];

const HR_OFFBOARDING_TYPES: WorkflowType[] = ["employee_offboarding"];
const IT_OFFBOARDING_TYPES: WorkflowType[] = ["employee_offboarding"];
const MASTER_OFFBOARDING_TYPES: WorkflowType[] = ["employee_offboarding"];

const EMPLOYEE_SUBSCRIPTION_TYPES: WorkflowType[] = ["employee_software_request"];

const MASTER_SUBSCRIPTION_TYPES: WorkflowType[] = [
  "employee_software_request",
  "new_subscription_request",
  "renewal_request",
];

const MASTER_LICENCE_TYPES: WorkflowType[] = ["license_assignment_request"];

export function canDirectCreateSoftwareRecords(roles: string[] = []): boolean {
  return roles.includes("master_admin");
}

export function getAllowedWorkflowTypes(
  moduleKey: string,
  roles: string[] = [],
  options?: { masterUsesWorkflow?: boolean }
): WorkflowType[] {
  if (moduleKey === "employees") {
    if (roles.includes("master_admin")) return MASTER_OFFBOARDING_TYPES;
    if (roles.includes("hr_admin")) return HR_OFFBOARDING_TYPES;
    if (roles.includes("it_admin")) return IT_OFFBOARDING_TYPES;
    return [];
  }
  if (!["subscriptions", "licences"].includes(moduleKey)) return [];
  if (roles.includes("master_admin")) {
    if (options?.masterUsesWorkflow) {
      return moduleKey === "subscriptions" ? MASTER_SUBSCRIPTION_TYPES : MASTER_LICENCE_TYPES;
    }
    return [];
  }
  if (roles.includes("employee")) {
    return moduleKey === "subscriptions" ? EMPLOYEE_SUBSCRIPTION_TYPES : [];
  }
  if (roles.includes("it_admin")) {
    return moduleKey === "subscriptions" ? IT_SUBSCRIPTION_TYPES : IT_LICENCE_TYPES;
  }
  if (roles.includes("hr_admin")) {
    return moduleKey === "licences" ? HR_LICENCE_TYPES : [];
  }
  return [];
}

export function canSubmitSoftwareWorkflow(
  moduleKey: string,
  roles: string[] = [],
  options?: { masterUsesWorkflow?: boolean }
): boolean {
  return getAllowedWorkflowTypes(moduleKey, roles, options).length > 0;
}

export function canSubmitWorkflowType(
  moduleKey: string,
  roles: string[] = [],
  workflowType: WorkflowType,
  options?: { masterUsesWorkflow?: boolean }
): boolean {
  return getAllowedWorkflowTypes(moduleKey, roles, options).includes(workflowType);
}

export function getDefaultWorkflowType(
  moduleKey: string,
  roles: string[] = [],
  options?: { masterUsesWorkflow?: boolean }
): WorkflowType | null {
  const allowed = getAllowedWorkflowTypes(moduleKey, roles, options);
  return allowed[0] ?? null;
}

export function canSubmitEmployeeSoftwareRequest(roles: string[] = []): boolean {
  return roles.includes("employee");
}

export function canManageToolRequestInbox(roles: string[] = []): boolean {
  return roles.includes("master_admin") || roles.includes("it_admin");
}

export function getPrimaryRoleLabel(roles: string[] = []): string {
  if (roles.includes("master_admin")) return "Master Admin";
  if (roles.includes("it_admin")) return "IT Admin";
  if (roles.includes("hr_admin")) return "HR Admin";
  if (roles.includes("finance")) return "Finance";
  if (roles.includes("line_manager")) return "Line Manager";
  if (roles.includes("employee")) return "Employee";
  if (roles.includes("auditor")) return "Auditor";
  return roles[0] || "User";
}

File: api/app/workflow_submission_permissions.py
"""Role-based permissions for subscription, licence, and tool request submissions."""

from __future__ import annotations

SOFTWARE_SUBMISSION_PERMISSIONS: dict[str, dict[str, set[str]]] = {
    "employee": {
        "subscriptions": {"employee_software_request"},
        "licences": set(),
    },
    "it_admin": {
        "subscriptions": {
            "new_subscription_request",
            "renewal_request",
        },
        "licences": {
            "license_assignment_request",
        },
    },
    "hr_admin": {
        "subscriptions": set(),
        "licences": {"license_assignment_request", "hr_onboarding_request"},
    },
}

TOOL_REQUEST_SUBMIT_ROLES = {"employee", "line_manager", "it_admin", "hr_admin", "master_admin"}
TOOL_REQUEST_INBOX_ROLES = {"it_admin", "master_admin"}


def _normalize_roles(actor_roles: list[str] | None) -> list[str]:
    if not actor_roles:
        return []
    return [str(role).strip() for role in actor_roles if str(role).strip()]


def validate_software_workflow_submission(
    actor_roles: list[str] | None,
    requested_module: str,
    workflow_type: str,
) -> None:
    if requested_module not in {"subscriptions", "licences"}:
        return

    roles = _normalize_roles(actor_roles)
    if "master_admin" in roles:
        return

    allowed: set[str] = set()
    for role in roles:
        module_permissions = SOFTWARE_SUBMISSION_PERMISSIONS.get(role, {})
        allowed.update(module_permissions.get(requested_module, set()))

    if not allowed:
        raise PermissionError(
            "You do not have permission to perform this action. "
            "Your role is not allowed to submit this request type."
        )

    if workflow_type not in allowed:
        raise PermissionError(
            f"You do not have permission to submit workflow type '{workflow_type}' on module '{requested_module}'."
        )


def validate_direct_software_create(actor_roles: list[str] | None, module: str) -> None:
    if module not in {"subscriptions", "licences"}:
        return
    roles = _normalize_roles(actor_roles)
    if "master_admin" not in roles:
        raise PermissionError(
            "You do not have permission to perform this action. "
            "Direct subscription and licence creation is restricted to Master Admin. Submit a workflow request instead."
        )


def validate_tool_request_submission(actor_roles: list[str] | None) -> None:
    roles = set(_normalize_roles(actor_roles))
    if not roles.intersection(TOOL_REQUEST_SUBMIT_ROLES):
        raise PermissionError(
            "You do not have permission to perform this action. "
            "Only employees and authorized staff may submit software requests."
        )


def validate_tool_request_inbox_action(actor_roles: list[str] | None) -> None:
    roles = set(_normalize_roles(actor_roles))
    if not roles.intersection(TOOL_REQUEST_INBOX_ROLES):
        raise PermissionError(
            "You do not have permission to perform this action. "
            "Only IT Admin and Master Admin may manage the tool request inbox."
        )


def validate_workflow_reject(actor_roles: list[str] | None, current_status: str) -> None:
    roles = set(_normalize_roles(actor_roles))
    status = str(current_status or "").lower()
    if "master_admin" in roles:
        return
    if "finance" in roles and status in {"line_manager_approved", "master_approved", "finance_approved", "submitted", "reopened"}:
        return
    if "line_manager" in roles and status in {"submitted", "reopened"}:
        return
    raise PermissionError("You do not have permission to reject this workflow at the current stage.")


def validate_workflow_reopen(actor_roles: list[str] | None) -> None:
    roles = set(_normalize_roles(actor_roles))
    if roles.intersection({"master_admin", "finance", "it_admin"}):
        return
    raise PermissionError("You do not have permission to reopen this workflow.")


def validate_workflow_request_info(actor_roles: list[str] | None, current_status: str) -> None:
    roles = set(_normalize_roles(actor_roles))
    status = str(current_status or "").lower()
    if "master_admin" in roles:
        return
    if "finance" in roles and status in {"line_manager_approved", "master_approved", "submitted", "reopened"}:
        return
    if "line_manager" in roles and status in {"submitted", "reopened"}:
        return
    if "it_admin" in roles and status in {"finance_approved", "submitted", "reopened"}:
        return
    raise PermissionError("You do not have permission to request more information at this stage.")


def validate_workflow_resubmit(actor_roles: list[str] | None, actor_email: str | None, request_row: dict) -> None:
    roles = set(_normalize_roles(actor_roles))
    status = str(request_row.get("status") or "").lower()
    if status != "info_requested":
        raise PermissionError("Only workflows awaiting more information can be resubmitted.")
    requester_email = str(request_row.get("requested_by_email") or "").strip().lower()
    actor = str(actor_email or "").strip().lower()
    if "master_admin" in roles:
        return
    if actor and requester_email and actor == requester_email:
        return
    raise PermissionError("You do not have permission to resubmit this workflow.")


def validate_workflow_procurement_complete(actor_roles: list[str] | None) -> None:
    roles = set(_normalize_roles(actor_roles))
    if roles.intersection({"master_admin", "it_admin"}):
        return
    raise PermissionError(
        "You do not have permission to perform this action. "
        "Only IT Admin or Master Admin can complete procurement and activation."
    )
